src/main/java/org/jenkinsci/plugins/workflow/graph/FlowScanner.java

```diff
@@ -0,0 +1,602 @@
+import org.jenkinsci.plugins.workflow.actions.StageAction;
+import org.jenkinsci.plugins.workflow.actions.WorkspaceAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+/**
+ * Generified algorithms for scanning pipeline flow graphs for information
+ * Supports a variety of algorithms for searching, and pluggable conditions
+ * Worth noting: predicates may be stateful here, and may see some or all of the nodes, depending on the scan method used.
+ *
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public class FlowScanner {
+
+    /**
+     * Create a predicate that will match on all FlowNodes having a specific action present
+     * @param actionClass Action class to look for
+     * @param <T> Action type
+     * @return Predicate that will match when FlowNode has the action given
+     */
+    @Nonnull
+    public static <T extends Action>  Predicate<FlowNode> nodeHasActionPredicate(@Nonnull final Class<T> actionClass) {
+        return new Predicate<FlowNode>() {
+            @Override
+            public boolean apply(FlowNode input) {
+                return (input != null && input.getAction(actionClass) != null);
+            }
+        };
+    }
+
+    // Default predicates
+    public static final Predicate<FlowNode> MATCH_HAS_LABEL = nodeHasActionPredicate(LabelAction.class);
+    public static final Predicate<FlowNode> MATCH_IS_STAGE = nodeHasActionPredicate(StageAction.class);
+    public static final Predicate<FlowNode> MATCH_HAS_WORKSPACE = nodeHasActionPredicate(WorkspaceAction.class);
+    public static final Predicate<FlowNode> MATCH_HAS_ERROR = nodeHasActionPredicate(ErrorAction.class);
+    public static final Predicate<FlowNode> MATCH_HAS_LOG = nodeHasActionPredicate(LogAction.class);
+    public static final Predicate<FlowNode> MATCH_BLOCK_START = (Predicate)Predicates.instanceOf(BlockStartNode.class);
+
+    public static Filterator<FlowNode> filterableEnclosingBlocks(FlowNode f) {
+        LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
+        scanner.setup(f);
+        return scanner.filter(MATCH_BLOCK_START);
+    }
+
+    /**
+     * Base class for flow scanners, which offers basic methods and stubs for algorithms
+     * Scanners store state internally, and are not thread-safe but are reusable
+     * Scans/analysis of graphs is implemented via internal iteration to allow reusing algorithm bodies
+     * However internal iteration has access to additional information
+     */
+    public static abstract class AbstractFlowScanner implements Iterable <FlowNode>, Filterator<FlowNode> {
```

Comment:
@jglick I figured this was coming: done and done, and it received some other refactoring and cleanup to boot.